designs/disruption-controls.md

```diff
@@ -0,0 +1,122 @@
+    Budgets []Budget `json:"budgets,omitempty" hash:"ignore"`
+    // ConsolidateAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Consolidation.
+    ConsolidateAfter *NillableDuration `json:"consolidateAfter" hash:"ignore"`
+    // ExpireAfter is a nillable duration, parsed as a metav1.Duration.
+    // Users can use "Never" to disable Expiration.
+    ExpireAfter *NillableDuration `json:"expireAfter" hash:"ignore"`
+    // ConsolidationPolicy determines how Karpenter will consider nodes
+    // as candidates for Consolidation.
+    // WhenEmpty uses the same behavior as v1alpha5 TTLSecondsAfterEmpty
+    // WhenUnderutilized uses the same behavior as v1alpha5.ConsolidationEnabled: true
+    ConsolidationPolicy string `json:"consolidationPolicy" hash:"ignore"`
+}
+// Budget specifies periods of times where Karpenter will restrict the
+// number of Node Claims that can be terminated at a time.
+// Unless specified, a budget is always active.
+type Budget struct {
+    // MaxUnavailable dictates how many Node Claims owned by this NodePool
+    // can be terminating at once.
+    // This only respects and considers nodes with a DeletionTimestamp set.
+    MaxUnavailable intstr.IntOrString `json:"maxUnavailable" hash:"ignore"`
+    // Crontab specifies when a budget begins being active.
+    // Crontab uses the same syntax as a Cronjob.
+    // And can support a TZ.
+    // "Minute Hour DayOfMonth Month DayOfWeek"
+    // This is required if Duration is set.
+    Crontab *string `json:"crontab,omitempty" hash:"ignore"`
+    // Duration determines how long a Budget is active since each Crontab hit.
+    // This is required if Crontab is set.
+    Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+```
+
+## Validation/Defaults
+
+For each `Budget`, `MaxUnavailable` is required, and must be non-negative. Users can disable scale down for a NodePool by setting this to `0`. Users must either omit both `Crontab` and `Duration` or set both of them, since `Crontab` and `Duration` are inherently linked. Omitting these two fields will be equivalent to an always active `Budget`. Users cannot define a seconds value in `Duration`, since the smallest denomination of time in upstream Crontabs are minutes.
+
+- Omitting the field `Budgets` will cause the field to be defaulted to one `Budget` with `MaxUnavailable: 10%`.
+- `ConsolidationPolicy` will be defaulted to `WhenUnderutilized`, with a `consolidateAfter` value of `15s`, which is the same value for Consolidation in v1alpha5.
+
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec:
+  disruption:
+    consolidationPolicy: WhenUnderutilized
+    consolidateAfter: 15s
+    expireAfter: 30d
+    budgets:
+    - maxUnavailable: 10%
+```
+Karpenter will not persist a default for the `Crontab` and `Duration` fields.
+
```

Comment:
I think this isn't cover but would love to see is how are we exposing these interactions either via metrics or events or both. I think Karpenter is already difficult enough to understand why it is making certain choices and this would make it even more difficult before any more complicated resource level or finer disruption budgets. In an ideal world, operators should be able to built a dashboard where at a glance they can understand why Node A isn't consolidated because of reasons X,Y,Z or why Node B will be consolidate in N minutes.

Reply:
Totally, I was hoping to keep this as a high level doc about the feature itself. We'll definitely be including metrics/events/statuses as we're always trying to make this better. As a start we could include how many disruptions are allowed at a given time for a NodePool into the status. Anything you had in mind here?

Reply:
I think if we can tied each deprovisioning method as a status to a nodeclaim and maybe mentioned that this machine isn't deprovisionable because of NDP would be very helpful